swarm/src/behaviour.rs

```diff
@@ -907,11 +609,11 @@ pub struct ConnectionEstablished<'a> {
 /// This event is always paired with an earlier
 /// [`FromSwarm::ConnectionEstablished`] with the same peer ID, connection ID
 /// and endpoint.
-pub struct ConnectionClosed<'a, Handler: IntoConnectionHandler> {
+pub struct ConnectionClosed<'a, Handler> {
     pub peer_id: PeerId,
     pub connection_id: ConnectionId,
     pub endpoint: &'a ConnectedPoint,
-    pub handler: <Handler as IntoConnectionHandler>::Handler,
+    pub handler: Handler,
```

Comment:
Out of scope for this PR, but I do wonder if we really need the `handler` field on `ConnectionClosed`. 
The `handler` property was added in #2191 together with adding it to `Dial`, `DialFailure` and `ListenFailure`. If I understand it correctly, the main motivation behind it was:

> it allows one to attach state to a dial request, thus not having to keep track of it within a NetworkBehaviour implementation 

Per https://github.com/libp2p/rust-libp2p/pull/3099#issuecomment-1313284181 attaching a state to a dial is not needed anymore. What's the use-case of still returning the handler in `ConnectionClosed`? As far as I can tell, none of our behaviours ever use it.

Motivation behind the question is that it would be great if we could get rid of the duplication between `SwarmEvent` and `FromSwarm`.
Until now the main difference was that `FromSwarm` also contains the handler, which should not be exposed in the `SwarmEvent`[^1]: https://github.com/libp2p/rust-libp2p/discussions/2423#discussioncomment-1941073. If we could remove it here, we might be able to do something like:
```rust
enum SwarmEvent<TBehaviourOutEvent> {
    Behaviour(TBehaviourOutEvent),
    FromSwarm(FromSwarmEvent)
}
```

[^1]: There are also some other differences; `FromSwarm` contains `ConnectionId`s, and the `NewExternalAddr` and `ExpiredExternalAddr` variants. But I don't see see a drawback in exposing them to the user.

Reply:
Related: https://github.com/libp2p/rust-libp2p/issues/3046